bluepysnap/circuit_ids.py

```diff
@@ -0,0 +1,222 @@
+
+    def get_populations(self, unique=False):
+        """Returns all population values from the circuit node IDs."""
+        if unique:
+            return self.index.levels[0].to_numpy()
+        return self.index.get_level_values(0).to_numpy()
+
+    def get_ids(self, unique=False):
+        """Returns all the ID values from the circuit node IDs."""
+        if unique:
+            return self.index.levels[1].to_numpy()
+        return self.index.get_level_values(1).to_numpy()
+
+    def copy(self):
+        """Copy a CircuitNodeIds."""
+        return CircuitNodeIds(self.index.copy())
+
+    def sort(self, inplace=False):
+        """Sort the CircuitNodeIds by population and then by ids.
+
+        Args:
+            inplace (bool): if set to True. Do the transformation inplace.
+
+        Notes:
+            sorting a CircuitNodeIds will gives better perf for the population extraction and the
+            data extraction from a sorted dataframe.
+        """
+        if not inplace:
+            return CircuitNodeIds(self.index, sort_index=True)
+        self.index = self.index.sortlevel()[0]
+        return None
+
+    def append(self, other, inplace=False):
+        """Append a CircuitNodeIds to the current one.
+
+        Args:
+            other (CircuitNodeIds): the other CircuitNodeIds to append to the current one.
+            inplace (bool): if set to True. Do the transformation inplace.
+        """
+        if not inplace:
+            return CircuitNodeIds(self.index.append(other.index))
+        self.index = self.index.append(other.index)
+        return None
+
+    def sample(self, sample_size, inplace=False):
+        """Sample a CircuitNodeIds.
+
+        Notes:
+            this function takes randomly ``sample_size`` values of a circuit node ids.
+
+        Args:
+            sample_size (int): the size of the sample. If the size of the sample is greater than
+                the size of the CircuitNodeIds then all ids are taken and shuffled.
+            inplace (bool): if set to True. Do the transformation inplace.
+        """
+        res = self if inplace else self.copy()
+        if len(res) != 0:
+            sample_size = sample_size if sample_size < len(res) else len(res)
+            res.index = res.index[np.random.choice(len(res), size=sample_size, replace=False)]
+        if not inplace:
+            return res
+        return None
+
+    def limit(self, limit_size, inplace=False):
+        """Limit the size of a CircuitNodeIds.
+
+        Notes:
+            this function takes the first ``limit_size`` values of a circuit node ids.
+
+        Args:
+            limit_size (int): the size of the sample. If the size of the sample is greater than
+                the size of the CircuitNodeIds then all ids are kept.
+            inplace (bool): if set to True. Do the transformation inplace.
+        """
+        res = self if inplace else self.copy()
+        res.index = res.index[0: limit_size]
```

Comment:
`limit` and `sample` are the same code except for the slice passed in `res.index[]`.

It may not be useful now, but for the future you can consider having a private function:

```python
def _slice_index(self, myslice, inplace=False):
        res = self if inplace else self.copy()
        res.index = res.index[myslice]
        if not inplace:
            return res
        return None
```

Then:
```python
     def limit(self, limit_size, inplace=False):
         return self._slice_index([0, limit_size], inplace=inplace)

     def sample(self, limit_size, inplace=False):
         # this one does not handle the len(res) == 0 but you see the idea
         return self._slice_index(np.random.choice(len(res), size=min(sample_size, len(res)), replace=False), 
                                  inplace=inplace)
```

Reply:
indeed. I totally forgot to factorize. 
done